README.md
## @balena/fetch

This is Balena's implementation of `fetch`, but others could find it useful too.

Features:
* [happy eyeballs](#happy-eyeballs-algorithm)
  * Attempt connections to all addresses provided [`getaddrinfo`](https://man7.org/linux/man-pages/man3/getaddrinfo.3.html), not just the first one
  * Attempt connection to both IPv6 and IPv4 concurrently for dual-stack hosts
  * Remember last successful connection type and try that first for next request
* browser/node safe
* ...more to come

This is meant to replace all naked calls to `http.request`, `fetch`, `node-fetch`, `@balena/request`, etc.


### Installation

```
npm i --save @balena/fetch
```

### Usage

```js
import fetch from '@balena/fetch';

const resp = await fetch('https://api.balena-cloud.com/ping');
console.log(await resp.text()) // <= OK
```

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v0.0.11
## (2021-09-26)

* refactor tests [Zane Hitchcox]
* add engines [Zane Hitchcox]

# v0.0.10
## (2021-09-26)

* Add abort controllers support timeout [Zane Hitchcox]

# v0.0.9
## (2021-09-26)

* explicitly create socket and call connect directly [Zane Hitchcox]

# v0.0.8
## (2021-09-26)

* skip timeout if cts fail [Zane Hitchcox]

# v0.0.7
## (2021-09-26)

* add happy eyes description [Zane Hitchcox]

# v0.0.6
## (2021-09-25)

* refactor getAddrInfo [Zane Hitchcox]
* move tests and zip addresses [Zane Hitchcox]

# v0.0.5
## (2021-09-24)

* add explanation what "happy eyeballs" is [Martin Rauscher]

# v0.0.4
## (2021-09-24)

* last used per hostname [Zane Hitchcox]

# v0.0.3
## (2021-09-24)

* remember last used, verbose logging, destroy correct sockets [Zane Hitchcox]

# v0.0.2
## (2021-09-24)

* add readme [Zane Hitchcox]


test/cli.ts
import { readdirSync } from 'fs';
import path from 'path';
import { name } from '../package.json';

async function cli() {
	console.log(name);
	process.env.NODE_DEBUG ??= '';
	process.env.NODE_ENV = name + '-test';
	const debugFlags: string[] = [];
	if (process.argv.includes('-d')) {
		debugFlags.push(name + '-debug');
	}
	if (process.argv.includes('-v')) {
		debugFlags.push(name + '-debug');
		debugFlags.push(name + '-verbose');
	}
	if (process.argv.includes('-m')) {
		debugFlags.push(name + '-manual');
	}
	process.env.NODE_DEBUG += debugFlags.join(',');

	// we must set up environment first
	const { run } = await import('./test');
	let files = process.argv
		.filter((file) => /\.ts$/.test(file))
		.map((f) => (path.isAbsolute(f) ? f : path.join(process.cwd(), f)));
	if (files.length < 2) {
		files = readdirSync(path.join(process.cwd(), 'src')).map((f) =>
			path.join(process.cwd(), 'src', f),
		);
	}

	for (const file of files) {
		await import(file).then(run);
	}

	if (process.argv.includes('-w')) {
		(async () => {
			const fs = await import('fs');
			for (const file of files) {
				fs.watchFile(file, async () => {
					console.log();
					delete require.cache[file];
					await import(file);
					await run();
				});
			}
		})();
	}
}

cli();


test/test.ts
import chalk from 'chalk';
import _expect from 'expect';
import { TESTING, MANUAL } from './constants';

// this will all be tree-shaken

// tslint:disable-next-line:no-empty
const noop = () => {};

type TestFn = {
	(...args: any): any | void;
	only?: boolean;
}

type TestData = {
	depth: number;
	name: string;
	fn: TestFn;
	only?: boolean;
	skip?: boolean;
}
const tests: Array<TestData> = [];

type Test = {
	(name: string, fn: TestFn): void;
	manual: (name: string, fn: TestFn) => void;
	skip: (...args: any[]) => any;
	only: (name: string, fn: TestFn) => any;
};

let depth = 0;
export const test = (TESTING
	? async (name: string, fn: TestFn) => {
			tests.push({depth, name, fn});
	  }
	: noop) as unknown as Test;

test.manual =
	TESTING && MANUAL
		? async (name: string, fn: TestFn) => {
				tests.push({depth, name, fn});
		  }
		: noop;

test.skip = (TESTING
	? (name: string, fn: TestFn) => {
			tests.push({depth, name, fn, skip: true});
	  }
	: noop) as unknown as Test;

test.only = (TESTING
	? (name: string, fn: TestFn) => {
			tests.push({depth, name, fn, only: true});
	  }
	: noop) as unknown as Test;


export const expect = TESTING ? _expect : (noop as unknown as typeof _expect);

// tslint:disable:no-conditional-assignment
export const run =
	(TESTING &&
		(async () => {
			if (tests.length === 0) {
				return;
			}

			const getOnlys = (tests: TestData[]) => {
				const onlys = tests.filter(test => test.only);
				return (onlys.length && onlys) || tests;
			}

			const batch = tests.slice();
			tests.length = 0;

			for (const test of getOnlys(batch)) {
				const {depth:_depth, name, fn, only, skip} = test;
				const indent = ' '.repeat(_depth);
				if (skip) {
					console.log(indent + chalk.yellow('○'), name);
					continue;
				}
				try {
					depth++;
					await fn();
					console.log(indent + chalk.green('✓'), name);
				} catch (error: any) {
					console.error(indent + chalk.red('☓'), name);
					console.error(chalk.red(error));
					console.error(error?.stack);
				}
				await run();
				depth--;
			}
		})) ||
	noop;


test/server.ts
import { Server } from 'http';
import { TESTING } from './constants';

// tslint:disable-next-line:no-empty
const noop = () => {};

export const createTestServer = (
	TESTING
		? async () => {
				const { createServer } = await import('http');
				return new Promise((pres, rej) => {
					const server = createServer((req, res) => {
						res.writeHead(200, {
							'Access-Control-Allow-Origin': '*',
						});
						res.write('It works!');
						res.end();
						req.destroy();
					})
						.listen(() => {
							// tslint:disable-next-line:ban-ts-ignore
							// @ts-ignore @types/node is wrong here
							const { port } = server.address();
							pres({ server, port });
						})
						.on('error', rej);
				});
		  }
		: noop
) as () => Promise<{ server: Server; port: number }>;


src/agent.ts
import { Agent as HttpAgent } from 'http';
import { Agent as HttpsAgent } from 'https';
import { BalenaRequestInit, happyEyeballs } from './happy-eyeballs';
import { URL } from 'url';
import * as net from 'net';

const agentCache = new Map<string, HttpsAgent|HttpAgent>();
export function getAgent(init: BalenaRequestInit) {
  const hash = getHash(init);
  // cache agents with same configuration to reduce connection overhead
  if (agentCache.has(hash)) {
    return agentCache.get(hash);
  }
  const agent = new class extends (init.secure ? HttpsAgent : HttpAgent) {
    createConnection = (url: URL, cb: (err: Error|undefined, socket?: net.Socket) => void) => {happyEyeballs(url, init, cb)};
  }
  agentCache.set(hash, agent);
  return agent;
}

// this can be eliminated once `node-fetch` accepts `createConnection` option (https://github.com/node-fetch/node-fetch/issues/1313)
const hashCache = new WeakMap<any, string>()
function getHash(item: any): string {
  switch (typeof item) {
    // we can cache objects and functions, in case people reuse the same
    // object or function in the options, so we don't have to create a hash again
    case 'object':
      if (hashCache.has(item)) {
        return hashCache.get(item)!;
      }
      // object has not been cached, so we need to create a new hash
      // this enables use to reuse http(s) agents which have the same
      // options for config, even if they are different objects
      let str = '{'
      const keys = typeof item.keys === 'function' ? item.keys() : Object.keys(item);
      for (const key of keys.sort()) {
        str+= `"${key}":${getHash(item[key])}`;
      }
      str += '}'
      // create hash of all stringified items
      let h = 0;
      for (const c of Buffer.from(str)) {
        h = ((h << 5) - h) + c | 0;
      }
      const hString = String(h);
      hashCache.set(item, hString);
      return hString
    case 'function':
      if (hashCache.has(item)) {
        return hashCache.get(item)!;
      }
      const id = String(Math.random());
      hashCache.set(item, id);
      return id!;
    case 'number':
    case 'boolean':
    case 'bigint':
    case 'symbol':
      return `"${item.toString()}"`;
    case 'undefined':
      return 'undefined'
  }
  // this should never be reached, but in case it does, bust cache
  return Math.random()+''
}

src/he-connect.ts
import * as net from 'net';
import * as tls from 'tls';
import * as dns from 'dns/promises';
import { debuglog, promisify } from 'util';
import { URL } from 'url';
import { Agent as HttpAgent } from 'http';
import { Agent as HttpsAgent } from 'https';
import { LookupAddress } from 'dns';

const NEXT_ADDR_DELAY = 300;

const debug = debuglog('@balena/fetch-debug');
const verbose = debuglog('@balena/fetch-verbose');
const sleep = promisify(setTimeout);
const lastSuccessful: {[host: string]: number | undefined} = {};

export const options = {
  agent: function(parsedURL: URL) {
    if (parsedURL.protocol === 'https:') {
      return new class extends HttpsAgent {
        createConnection = createConnection
      }
    } else {
      return new class extends HttpAgent {
        createConnection = createConnection
      }
    }
  }
}

function createConnection(url:URL, cb: (err: Error | undefined, socket?: net.Socket) => void) {
  debug('Connecting to', url.hostname);
  happyEyeballs(url, cb)
}

export async function happyEyeballs(url: URL, cb: (err: Error | undefined, socket?: net.Socket) => void) {
  const {hostname, protocol} = url;
  const preferred = lastSuccessful[url.hostname];
  const lookups = await dns.lookup(hostname, {
    verbatim: true,
    family: 0,
    all: true,
  });
  if (!lookups.length) {
    throw new Error(`Could not resolve host, ${hostname}`);
  }
  const addrs = await getAddrInfo(lookups, preferred);
  const port = url.port.length ? Number(url.port) : (protocol === "https:" ? 443 : 80);

  let numErrs = 0;
  let firstError: Error;
  let ctFound = false;
  const sockets: net.Socket[] = [];
  for (const tuple of addrs) {
    if (ctFound) {
      break;
    }
    for (const host of tuple) {
      debug(`Trying ${host}...`);
      const socket = (protocol === 'https:' ? tls : net as unknown as typeof tls).connect({
        host,
        port,
        servername: hostname,
      }).on('connect', () => {
        if (ctFound) {
          socket.destroy();
          return;
        }
        cb(undefined, socket);
        ctFound = true;
        for (const s of sockets) {
          if (s !== socket) {
            debug('Destroying', s.remoteAddress);
            s.destroy();
          }
        }
        if (net.isIPv4(host)) {
          lastSuccessful[hostname]  = 4;
        } else {
          lastSuccessful[hostname] = 6;
        }
        debug('Connected to', socket.remoteAddress);
      }).on('error', (err: Error) => {
        if (numErrs === 0) {
          firstError = err;
        }
        numErrs++;
        if (numErrs === lookups.length) {
          cb(firstError);
        }
      })
      sockets.push(socket);
    }
    // give each connection 300 ms to connect before trying next one
    await sleep(NEXT_ADDR_DELAY);
  }
}

export function*getAddrInfo(lookups: LookupAddress[], preferred: number | undefined): Iterable<string[]> {
  let next = preferred ?? lookups[0].family;
  const queue: string[] = [];
  for (const {address, family} of lookups) {
    if (family === next) {
      // if there is no preferred address, give tuples of IPv6 and IPv4 addresses
      // try both at the same time.
      if (!preferred) {
        if (queue.length) {
          // queue will always hold the opposite of what we're looking for
          // If there is an item on the queue, we have found a pair.
          yield [address, queue.shift() as string]
        } else {
          // if there was no queue item, create one, and switch `next`
          // to the opposite family.
          queue.push(address);
          next = family === 6 ? 4 : 6;
        }
      } else {
        // we've seen this host before, so we just return one at a time
        // we still alternate though, in case network has changed
        yield [address];
        if (queue.length) {
          // If there was a queue, that means it holds the opposite family
          // so go ahead and yield that item too
          yield [queue.shift()] as string[];
        } else {
          // There was no queue item. So, we queue this item and
          // switch `next` to the opposite family
          queue.push(address);
          next = family === 6 ? 4 : 6;
        }
      }
    } else {
      queue.push(address)
    }
  }
  for (const addr of queue) {
    yield [addr];
  }
}

src/abort-controller.ts
import { addAbortSignal, EventEmitter } from "stream";

interface EventListener {
    (evt: Event): void;
}

interface EventListenerObject {
    handleEvent(object: Event): void;
}

type EventListenerOrEventListenerObject = EventListener | EventListenerObject;

interface EventListenerOptions {
    capture?: boolean;
}

interface AbortSignalEventMap {
    "abort": Event;
}

/** EventTarget is a DOM interface implemented by objects that can receive events and may have listeners for them. */
interface EventTarget {
    /**
     * Appends an event listener for events whose type attribute value is type. The callback argument sets the callback that will be invoked when the event is dispatched.
     *
     * The options argument sets listener-specific options. For compatibility this can be a boolean, in which case the method behaves exactly as if the value was specified as options's capture.
     *
     * When set to true, options's capture prevents callback from being invoked when the event's eventPhase attribute value is BUBBLING_PHASE. When false (or not present), callback will not be invoked when event's eventPhase attribute value is CAPTURING_PHASE. Either way, callback will be invoked if event's eventPhase attribute value is AT_TARGET.
     *
     * When set to true, options's passive indicates that the callback will not cancel the event by invoking preventDefault(). This is used to enable performance optimizations described in § 2.8 Observing event listeners.
     *
     * When set to true, options's once indicates that the callback will only be invoked once after which the event listener will be removed.
     *
     * If an AbortSignal is passed for options's signal, then the event listener will be removed when signal is aborted.
     *
     * The event listener is appended to target's event listener list and is not appended if it has the same type, callback, and capture.
     */
    addEventListener(type: string, callback: EventListenerOrEventListenerObject | null, options?: AddEventListenerOptions | boolean): void;
    /**
     * Dispatches a synthetic event event to target and returns true if either event's cancelable attribute value is false or its preventDefault() method was not invoked, and false otherwise.
     */
    dispatchEvent(event: Event): boolean;
    /**
     * Removes the event listener in target's event listener list with the same type, callback, and options.
     */
    removeEventListener(type: string, callback: EventListenerOrEventListenerObject | null, options?: EventListenerOptions | boolean): void;
}

/** A signal object that allows you to communicate with a DOM request (such as a Fetch) and abort it if required via an AbortController object. */
interface AS extends EventEmitter,EventTarget {
    /**
     * Returns true if this AbortSignal's AbortController has signaled to abort, and false otherwise.
     */
    readonly aborted: boolean;
    onabort: ((this: AbortSignal, ev: Event) => any) | null;
    addEventListener<K extends keyof AbortSignalEventMap>(type: K, listener: (this: AbortSignal, ev: AbortSignalEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
    addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;
    removeEventListener<K extends keyof AbortSignalEventMap>(type: K, listener: (this: AbortSignal, ev: AbortSignalEventMap[K]) => any, options?: boolean | EventListenerOptions): void;
    removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void;
}

interface AddEventListenerOptions extends EventListenerOptions {
    once?: boolean;
    passive?: boolean;
    signal?: AbortSignal;
}

export class AbortController {
  public readonly signal = new addAbortSignal();
}

export class AbortSignal extends EventEmitter implements AS {
    public onabort = null;
    public addEventListener<K extends keyof AbortSignalEventMap>(type: K, listener: (this: AbortSignal, ev: AbortSignalEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
    public addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;
    public addEventListener(type, listener) {
      this.off(type, listener);
    }
    public removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions) {
      if (typeof options === 'boolean') {
        options =
      }
      this.on(type, listener);
    }
    // addEventListener<K extends keyof AbortSignalEventMap>(type: K, listener: (this: AbortSignal, ev: AbortSignalEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
    // addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;
    // removeEventListener<K extends keyof AbortSignalEventMap>(type: K, listener: (this: AbortSignal, ev: AbortSignalEventMap[K]) => any, options?: boolean | EventListenerOptions): void;
    // removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void;
}

// // node-style abort controller
// export class AbortController extends EventEmitter {
//   private _aborted = false;
//   public get aborted() {
//     return this._aborted;
//   }
//   constructor() {

//   }
//   abort() {
//   }
//   toString() {
//     return '[object AbortController]';
//   }
// }

// export default AbortController;

// if (typeof Symbol !== 'undefined' && Symbol.toStringTag) {
//   // These are necessary to make sure that we get correct output for:
//   // Object.prototype.toString.call(new AbortController())
//   AbortController.prototype[Symbol.toStringTag] = 'AbortController';
//   AbortSignal.prototype[Symbol.toStringTag] = 'AbortSignal';
// }

src/happy-eyeballs.ts
import * as net from 'net';
import * as tls from 'tls';
import * as dns from 'dns';
import { debuglog, promisify } from 'util';
import { URL } from 'url';
import { LookupAddress, LookupOptions } from 'dns';
import { RequestInit as NFRequestInit } from 'node-fetch'
import AbortError from './abort-error';
import { AbortController, AbortSignal } from 'abort-controller';
import {once} from 'events';

const dnsLookup = promisify(dns.lookup);

const DEFAULT_NEXT_ADDR_DELAY = 300;
const DEFAULT_LOOKUP_OPTIONS = Object.freeze({
  verbatim: true,
  family: 0,
  all: true,
});

const debug = debuglog('@balena/fetch-debug');
// const verbose = debuglog('@balena/fetch-verbose');

// hash of hosts and last associated connection family
const familyCache = new Map<string, number>();

export type BalenaRequestInit = NFRequestInit & RequestInit & {
  secure: boolean;
  delay?: number; // ms delay between requests
  lookup?: (hostname: string, options: LookupOptions) => Promise<LookupAddress | LookupAddress[]>;
  family?: number;
  hints?: number;
  signal?: AbortSignal;
  timeout?: number;
};

export async function happyEyeballs(url: URL, init: BalenaRequestInit, cb: (err: Error | undefined, socket?: net.Socket) => void) {
  const {hostname} = url;
  debug('Connecting to', hostname);

  const lookups = (await (init.lookup || dnsLookup)(hostname, {
    ...DEFAULT_LOOKUP_OPTIONS,
    family: init.family || DEFAULT_LOOKUP_OPTIONS.family,
    hints: init.hints,
  })) as LookupAddress[];
  if (!lookups.length) {
    throw new Error(`Could not resolve host, ${hostname}`);
  }

  const sockets = new Map<string, net.Socket>();
  init.signal?.addEventListener('abort', () => {
    debug('Received abort signal, destroying all sockets.');
    for (const socket of sockets.values()) {
      socket.destroy();
    }
    cb(new AbortError);
  })
  const port = url.port.length ? Number(url.port) : (init.secure ? 443 : 80);

  let trying = lookups.length;
  let err: Error;
  function onError(this: net.Socket, _err:any) {
    if (init.signal?.aborted){
      return;
    }
    debug('Got error', _err)

    // Only use the value of the first error, as that was the one most likely to succeed
    if (!err){
      err = _err
    }

    this.destroy();

    debug('trying', trying)
    if (!--trying) {
      if (Array.from(sockets.values()).every(s => s.destroyed)) {
        debug('all sockets destroyed');
      }
      debug('All addresses failed')
      return cb(err)
    }
    debug('More addresses to try, continuing...');
  }

  let ctFound = false;
  function onConnect(this: net.Socket) {
    debug('Connected to', this.remoteAddress);
    ctFound = true;
    for (const s of sockets.values()) {
      if (s !== this) {
        debug('Destroying', s.remoteAddress);
        s.destroy();
      }
    }
    // save last successful connection family for future reference
    familyCache.set(hostname, net.isIP(this.remoteAddress!));
    cb(undefined, this);
  }

  for (const batch of zip(lookups, familyCache.get(hostname))) {
    debug('batch', batch);
    if (ctFound || init.signal?.aborted) {
      return;
    }

    for (const addr of batch) {
      debug(`Trying ${addr}...`);
      const socket = (new net.Socket())
        .connect({
          host: addr,
          port,
        });
      if(init.timeout) {
        debug('Setting timeout, ' + init.timeout);
        socket.setTimeout(init.timeout);
        socket.on('timeout', function(this: net.Socket) {
          --trying;
          this.destroy();
          if (!trying) {
            if (Array.from(sockets.values()).every(s => !s.destroyed)) {
              cb(err || new Error('All connection attempts to ' + hostname + ' timed out.'))
            }
          }
          debug(`Request to ${addr} timed out...`)
        })
      }
      const ct = init.secure ? tls.connect({
        socket,
        servername: hostname,
        timeout: init.timeout,
      }) : socket;
      sockets.set(addr, ct);
      ct
        .on('connect', onConnect)
        .on('error', onError);
    }

    // abort the delay if all sockets close before the delay runs out,
    const {abort:_abort, signal} = new AbortController();
    const abort = () => {
      debug('Aborting delay promise');
      _abort();
    };
    init.signal?.addEventListener('abort', abort);
    Promise.all(batch
      .map(addr => sockets.get(addr))
      .map(sck => Promise.race([
        once(sck!, 'close', {signal}),
        once(sck!, 'error', {signal}),
      ]))
    )
      .then(() => abort())
      .catch(() => {});

    try {
      // give each connection 300 ms to connect before trying next one
      await wait(init.delay || DEFAULT_NEXT_ADDR_DELAY, signal);
    } catch (err: any) {
      if (err?.name !== 'AbortError') {
        throw err;
      }
    }
    init.signal?.removeEventListener('abort', abort);
  }
}

// this function follows the happy eyeballs 2 algorithm: https://datatracker.ietf.org/doc/html/rfc8305
export function*zip(lookups: LookupAddress[], init?: number): Iterable<string[]> {
  // `init` is the cached value of the family of the last successful connection
  // or `undefined` if no successful connections have been made to this host

  // `next` is the next address family we are looking for
  let next = init;

  // queue of addresses not matching `next`
  const queue: string[] = [];

  for (const {address, family} of lookups) {
    if (family === next || !next) {
      if (init) {
        // we've seen this host before, so just try this connection first
        yield [address];
        // set init to 0, so dual-stack addresses will be returned if the first one fails
        init = 0;
      } else {
        // the cached address family didn't connect in time, or there was no cached address family
        // so now yield pairs of both families
        if (queue.length) {
          // If there is an item on the queue, we have found a pair of mixed families.
          yield [address, queue.shift()!]
        } else {
          // queue was empty, so queue this item and switch the family we're looking for
          queue.push(address);
          next = family === 6 ? 4 : 6;
        }
      }
    } else {
      queue.push(address)
    }
  }
  // The leftover from the queue are all from the same family
  // so, just return a single-value array to try one at a time
  for (const addr of queue) {
    yield [addr];
  }
}

// We may want to abort a wait to keep the wait handle from keeping the process open
const wait = async (ms: number, signal?: AbortSignal) => {
  return signal?.aborted ?
    Promise.reject(new AbortError()) :
    new Promise<void>((res, rej) => {
      const onAbort = () => {
        clearTimeout(timeout);
        rej(new AbortError());
      };
      const timeout = setTimeout(() => {
        signal?.removeEventListener('abort', onAbort);
        res();
      }, ms);
      signal?.addEventListener('abort', onAbort);
  });
}
